### PR TITLE
Macros for creating solutions and workups

### PR DESCRIPTION
**Pull request by brilee**

### CODE REVIEW

**Comment by brilee** [comment]
@skearnes tests passing now

File: ord_schema/macros/workups.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    for component in workup.input.components:
+        component.reaction_role = reaction_pb2.ReactionRole.WORKUP
+    return workup
+
+
+def separate_phases(keep_phase: str):
+    return reaction_pb2.ReactionWorkup(type='EXTRACTION', keep_phase=keep_phase)
+
+
+def drying_agent(agent_smiles: str):
```
Comment (skearnes) [inline ord_schema/macros/workups.py:68]: please add type annotations and docstrings to these functions

File: ord_schema/macros/solutions.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        concentration = CONCENTRATION_RESOLVER.resolve(concentration)
+
+    if volume is not None and concentration is not None:
+        solute_amount = units.compute_solute_quantity(volume, concentration)
+    else:
+        solute_amount = None
+
+    output_compounds = []
+    solvent_pb = reaction_pb2.Compound()
+    solvent_pb.identifiers.add(value=solvent_smiles, type='SMILES')
```
Comment (skearnes) [inline ord_schema/macros/solutions.py:43]: what happens if solvent_smiles is None?

File: ord_schema/macros/solutions.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    if solute_smiles:
+        solute_pb = reaction_pb2.Compound()
+        solute_pb.identifiers.add(value=solute_smiles, type='SMILES')
+        if solute_amount is not None:
+            solute_pb.amount.MergeFrom(solute_amount)
+        output_compounds.append(solute_pb)
+    return output_compounds
+
+
+def BRINE(volume=None):
```
Comment (skearnes) [inline ord_schema/macros/solutions.py:59]: why is this named like a constant?

File: ord_schema/units.py [hunk 4]
```diff
@@ -237,3 +237,38 @@ def format_message(message: ord_schema.UnitMessage) -> Optional[str]:
         txt += f'(± {message.precision:.7g}) '
     txt += _UNIT_SYNONYMS[type(message)][message.units][0]
     return txt
+
+
+def compute_solute_quantity(
```
Comment (skearnes) [inline ord_schema/units.py:242]: this might require some updates after #610 goes in

File: ord_schema/macros/workups.py (hunk 1 above)
Comment (brilee) [reply ord_schema/macros/workups.py]: Done

File: ord_schema/macros/solutions.py (hunk 2 above)
Comment (brilee) [reply ord_schema/macros/solutions.py:68]: Changed type annotation to non-optional.

File: ord_schema/macros/solutions.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+CONCENTRATION_RESOLVER = units.UnitResolver(units.CONCENTRATION_UNIT_SYNONYMS,
+                                            forbidden_units={})
+
+
+def simple_solution(
+    solvent_smiles=None,
+    solute_smiles=None,
+    volume: Union[None, str, reaction_pb2.Volume] = None,
+    concentration: Union[None, str, reaction_pb2.Concentration] = None
+) -> List[reaction_pb2.Compound]:
```
[the remark this answers is not in the document]
Comment (brilee) [reply ord_schema/macros/solutions.py:31]: done

File: ord_schema/macros/solutions.py (hunk 3 above)
Comment (brilee) [reply ord_schema/macros/solutions.py]: changed to lowercase

File: ord_schema/units.py (hunk 4 above)
Comment (brilee) [reply ord_schema/units.py:242]: This was okay but the brine() macro needed updates to use the new SATURATED concept.

**Comment by brilee** [comment]
Looks like pylint picked up two new rules we are now violating: W1514 (open() must be used with explicit encoding) and C0209 - use f-strings instead of .format(). what do you want to do about these? The first seems fix-worthy, the second kind of annoying.

File: ord_schema/message_helpers_test.py [hunk 6]
```diff
@@ -380,8 +380,8 @@ def test_set_solute_moles(self):
         message_helpers.set_solute_moles(solute, [solvent3],
                                          '3 mM',
                                          overwrite=True)
-        self.assertEqual(solute.amount.moles,
-                         reaction_pb2.Moles(units='NANOMOLE', value=225))
+        self.assertEqual(solute.amount.moles.units, reaction_pb2.Moles.NANOMOLE)
```
Comment (skearnes) [inline ord_schema/message_helpers_test.py:383]: why did you change this test? proto equality seems better than field equality

File: ord_schema/units.py [hunk 7]
```diff
@@ -237,3 +237,38 @@ def format_message(message: ord_schema.UnitMessage) -> Optional[str]:
...
+        volume: reaction_pb2.Volume,
+        concentration: reaction_pb2.Concentration) -> reaction_pb2.Amount:
+    """Computes the quantity of a solute, given volume and concentration."""
+    volume_conversion = {
```
Comment (skearnes) [inline ord_schema/units.py:246]: move this out as a CONSTANT?

File: ord_schema/units.py [hunk 8]
```diff
@@ -237,3 +237,38 @@ def format_message(message: ord_schema.UnitMessage) -> Optional[str]:
...
+        volume.LITER: 1,
+        volume.MILLILITER: 1e-3,
+        volume.MICROLITER: 1e-6,
+        volume.NANOLITER: 1e-9,
+    }[volume.units]
+    concentration_conversion = {
```
Comment (skearnes) [inline ord_schema/units.py:252]: same here

File: ord_schema/macros/solutions_test.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+from absl.testing import absltest
+from absl.testing import parameterized
+
+from ord_schema.proto import reaction_pb2
+from ord_schema.macros import solutions
+from ord_schema import validations
+
+
+class SolutionsTest(parameterized.TestCase, absltest.TestCase):
```
Comment (skearnes) [inline ord_schema/macros/solutions_test.py:25]: Add some tests with expected output?

File: ord_schema/macros/solutions.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    if isinstance(concentration, str):
+        concentration = CONCENTRATION_RESOLVER.resolve(concentration)
+
+    if volume is not None and concentration is not None:
+        solute_amount = units.compute_solute_quantity(volume, concentration)
+    elif saturated:
+        solute_amount = reaction_pb2.Amount(
+            unmeasured=reaction_pb2.UnmeasuredAmount(type='SATURATED'))
+    else:
+        solute_amount = None
```
Comment (skearnes) [inline ord_schema/macros/solutions.py:64]: when would this happen/be allowed?

**Review by skearnes — APPROVED** [review]

**Comment by skearnes** [comment]
> Looks like pylint picked up two new rules we are now violating: W1514 (open() must be used with explicit encoding) and C0209 - use f-strings instead of .format(). what do you want to do about these? The first seems fix-worthy, the second kind of annoying.

Those both seem like things that should be fixed. If there are places where f-strings don't make sense, we can disable them inline.

File: ord_schema/message_helpers_test.py (hunk 6 above)
Comment (brilee) [reply ord_schema/message_helpers_test.py:383]: I forgot why, too... rerunning the old code shows this error. It seems to be due to a subtle reording of floating point ops / units rescaling of nanomoles and whatnot. 

Traceback (most recent call last):
  File "./ord_schema/message_helpers_test.py", line 386, in test_set_solute_moles
    reaction_pb2.Moles(units='NANOMOLE', value=225))
AssertionError: value: 225.00001525878906
units: NANOMOLE
 != value: 225.0
units: NANOMOLE

File: ord_schema/macros/solutions_test.py (hunk 9 above)
Comment (brilee) [reply ord_schema/macros/solutions_test.py:26]: done

**Comment by brilee** [comment]
added tests / resolve comments

**Comment by brilee** [comment]
currently debugging the test_notebooks failure.

Traceback (most recent call last):
  File "/usr/share/miniconda/lib/python3.7/site-packages/traitlets/traitlets.py", line 537, in get
    value = obj._trait_values[self.name]
KeyError: 'kernel_dirs'

...

  File "/usr/share/miniconda/lib/python3.7/os.py", line 223, in makedirs
    mkdir(name, mode)
FileExistsError: [Errno 17] File exists: '/home/runner/.ipython'


I believe this is caused by Treon running multiple jupyter kernels at once, and there is likely a race condition where each kernel sees that the lazily populated kernel_dirs traitlet doesn't exist and then tries to create it...

**Comment by brilee** [comment]
this recent commit seems like alikely culprit: https://github.com/ipython/ipython/commit/7a20999bb2097a781b81e3098ba3e379cdd6b14e#diff-3adb2b67c2d8a4ccf260824b6391483e94429fcd20e547c4729bb5a55855eae9

**Comment by brilee** [comment]
in the meantime, I reran the test and it seems to have randomly not triggered the race condition.
